### PR TITLE
Fix push notifications where channel is set to all activity

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
This should fix the push notifications not being sent when: 
* The Notification global settings are set to something different than `all activity` and the channel is set to `all activity`
* The Notification global settings are set to something different than `never` and the channel setting is set to something else

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-3193

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: app/notification.go [hunk 1]
```diff
@@ -102,9 +102,10 @@ func SendNotifications(post *model.Post, team *model.Team, channel *model.Channe
 
 		// find which users in the channel are set up to always receive mobile notifications
 		for _, profile := range profileMap {
-			if profile.NotifyProps[model.PUSH_NOTIFY_PROP] == model.USER_NOTIFY_ALL &&
+			if (profile.NotifyProps[model.PUSH_NOTIFY_PROP] == model.USER_NOTIFY_ALL &&
 				(post.UserId != profile.Id || post.Props["from_webhook"] == "true") &&
-				!post.IsSystemMessage() {
+				!post.IsSystemMessage()) ||
```
Comment (crspeller) [inline app/notification.go]: I think the `(post.UserId != profile.Id || post.Props["from_webhook"] == "true") && !post.IsSystemMessage()` Shouldn't be inside the brackets. Ex you never want to recieve pushes for system messages.

Comment (enahum) [inline app/notification.go]: got it, totally right I'll fix it now

File: app/status.go [hunk 2]
```diff
@@ -236,10 +236,12 @@ func IsUserAway(lastActivityAt int64) bool {
 	return model.GetMillis()-lastActivityAt >= *utils.Cfg.TeamSettings.UserStatusAwayTimeout*1000
 }
 
-func DoesStatusAllowPushNotification(user *model.User, status *model.Status, channelId string) bool {
+func DoesStatusAllowPushNotification(user *model.User, channelNotifyProps model.StringMap, status *model.Status, channelId string) bool {
 	props := user.NotifyProps
 
-	if props["push"] == "none" {
+	if props[model.PUSH_NOTIFY_PROP] == model.USER_NOTIFY_NONE &&
```
Comment (crspeller) [inline app/status.go]: Actually This logic my be wrong too. If the global notify is not none but the channel is none it won't return false.

File: app/status.go [hunk 3]
```diff
@@ -236,10 +236,21 @@ func IsUserAway(lastActivityAt int64) bool {
 	return model.GetMillis()-lastActivityAt >= *utils.Cfg.TeamSettings.UserStatusAwayTimeout*1000
 }
 
-func DoesStatusAllowPushNotification(user *model.User, status *model.Status, channelId string) bool {
+func DoesStatusAllowPushNotification(user *model.User, channelNotifyProps model.StringMap, wasMentioned bool, status *model.Status, channelId string) bool {
```
Comment (crspeller) [inline app/status.go]: Maybe rename to `ShouldSendPushNotification`? It's checking both status and props. You could even split those two responsibilities into different functions.

File: app/status.go [hunk 4]
```diff
@@ -236,10 +236,21 @@ func IsUserAway(lastActivityAt int64) bool {
...
+	ok := false
+	if channelNotify, ok = channelNotifyProps[model.PUSH_NOTIFY_PROP]; ok {
+		if channelNotify == model.USER_NOTIFY_NONE {
+			return false
+		} else if channelNotify == model.CHANNEL_NOTIFY_MENTION && !wasMentioned {
+			return false
+		}
+	}
+
+	if props[model.PUSH_NOTIFY_PROP] == model.USER_NOTIFY_NONE &&
```
Comment (crspeller) [inline app/status.go]: Shouldn't there be wasMentioned logic like above? So this currently would not catch the case where the default is CHANNEL_NOTIFY_MENTION and the person was not mentioned.

Comment (enahum) [inline app/status.go]: Then if that's the case the function won't be called

File: app/status.go (hunk 3 above)
Comment (enahum) [inline app/status.go]: I can rename it and split it, not problem

**Comment by enahum** [comment]
@crspeller have a look and let me know

File: app/notification.go [hunk 5]
```diff
@@ -742,3 +735,48 @@ func GetMentionKeywordsInChannel(profiles map[string]*model.User) map[string][]s
 
 	return keywords
 }
+
+func ShouldSendPushNotification(user *model.User, channelNotifyProps model.StringMap, wasMentioned bool, status *model.Status, post *model.Post) bool {
+	return DoesNotifyPropsAllowPushNotification(user, channelNotifyProps, post, wasMentioned) &&
+		DoesStatusAllowPushNotification(user.NotifyProps, status, post.ChannelId)
+}
+
+func DoesNotifyPropsAllowPushNotification(user *model.User, channelNotifyProps model.StringMap, post *model.Post, wasMentioned bool) bool {
```
Comment (crspeller) [inline app/notification.go]: I don't have confidence in this function. Can we write a unit test to test every case so we know it works?

File: app/notification.go [hunk 6]
```diff
@@ -742,3 +735,48 @@ func GetMentionKeywordsInChannel(profiles map[string]*model.User) map[string][]s
...
+
+	if userNotify == model.USER_NOTIFY_NONE &&
+		(!ok || channelNotify == model.CHANNEL_NOTIFY_DEFAULT) {
+		return false
+	}
+
+	return true
+}
+
+func DoesStatusAllowPushNotification(userNotifyProps model.StringMap, status *model.Status, channelId string) bool {
```
Comment (crspeller) [inline app/notification.go]: Unit test wouldn't hurt for this one.

File: app/notification.go (hunk 5 above)
Comment (enahum) [inline app/notification.go]: will do some time today

File: app/notification.go (hunk 6 above)
Comment (enahum) [inline app/notification.go]: same for this one, although this one is basically the same one that is currently in

**Comment by enahum** [comment]
@crspeller unit tests added

**Pull request merged**